pyomo/contrib/community_detection/community_graph.py

```diff
@@ -0,0 +1,343 @@
+                collapsed_model_graph.add_node(node)
+
+    elif type_of_graph == 'v':  # Variable node graph - collapse the bipartite graph into a variable node graph
+
+        for node in bipartite_model_graph.nodes():
+
+            # If the node represents a constraint (or objective)
+            if node in constraint_variable_map:
+
+                # Create list of variables in the constraint equation
+                connected_variables = constraint_variable_map[node]
+
+                # Create all possible two-node combinations from connected_variables; in other words, create a list
+                # of all the edges that need to be created between variables based on connected_variables
+                edges_between_variables = list(combinations(sorted(connected_variables), 2))
+
+                # Update edge_weight_dict or edge_set based on the determined edges_between_nodes
+                if weighted_graph:
+                    new_edge_weights = {edge: edge_weight_dict.get(edge, 0) + 1 for edge in edges_between_variables}
+                    edge_weight_dict.update(new_edge_weights)
+                else:
+                    edge_set.update(edges_between_variables)
+
+            # If the node is not a constraint/objective, then it is a variable and we want it as
+            # a node in our variable node graph
+            else:
+                collapsed_model_graph.add_node(node)
+
+    # Now, using edge_weight_dict or edge_set (based on if the user wants a weighted graph or an unweighted graph,
+    # respectively), the NetworkX graph (collapsed_model_graph) will be updated with all of the edges determined above
+    if weighted_graph:
+
+        # Add edges to collapsed_model_graph (the order in which edges are added can affect community detection, so
+        # sorting prevents any unpredictable changes)
+        collapsed_model_graph.add_edges_from(sorted(edge_weight_dict))
+
+        # Iterate through the edges in edge_weight_dict and add them to collapsed_model_graph
+        for edge in edge_weight_dict:
+            node_one, node_two = edge[0], edge[1]
+            collapsed_model_graph[node_one][node_two]['weight'] = edge_weight_dict[edge]
+
+    else:
+        # Add edges to collapsed_model_graph (the order in which edges are added can affect community detection, so
+        # sorting prevents any unpredictable changes)
+        collapsed_model_graph.add_edges_from(sorted(edge_set))
+
+    # Log important information with the following logger function
+    _event_log(model, collapsed_model_graph, set(constraint_variable_map), type_of_graph, with_objective)
+
+    # Return the collapsed NetworkX graph, the dictionary of node numbers mapped to their respective Pyomo
+    # components, and the map of constraints to the variables they contain
+    return collapsed_model_graph, number_component_map, constraint_variable_map
+
+
+def _event_log(model, model_graph, constraint_set, type_of_graph, with_objective):
```

Comment:
I would move this into a separate utility module, since it's not business logic.